serverless/email/status/main.go

```diff
@@ -0,0 +1,150 @@
+)
+
+var (
+	// setup context/logger
+	ctx, logger = logging.SetupLoggerWithLevel(context.Background(), zerolog.InfoLevel)
+
+	// tables
+	statusTable = aws.String("status")
+
+	// clients
+	dynamoClient *dynamodb.Client
+)
+
+func init() {
+	// setup base aws config
+	config, err := appaws.BaseAWSConfig(ctx, logger)
+	if err != nil {
+		panic("failed to create aws config")
+	}
+	// setup dynamodb client
+	dynamoClient = dynamodb.NewFromConfig(config)
+}
+
+// We use event publishing to receive delivery status notifications via SNS
+// https://docs.aws.amazon.com/ses/latest/dg/event-publishing-retrieving-sns-contents.html
+type sesNotification struct {
+	EventType     string         `json:"eventType"`
+	Mail          *mail          `json:"mail"`
+	Bounce        *bounce        `json:"bounce"`
+	Complaint     *complaint     `json:"complaint"`
+	Delivery      *delivery      `json:"delivery"`
+	DeliveryDelay *deliveryDelay `json:"deliveryDelay"`
+}
+
+type mail struct {
+	MessageID string                `json:"messageId"`
+	Tags      []sestypes.MessageTag `json:"tags"`
+}
+
+type bounce struct {
+	Timestamp time.Time `json:"timestamp"`
+}
+
+type complaint struct {
+	Timestamp time.Time `json:"timestamp"`
+}
+
+type delivery struct {
+	Timestamp time.Time `json:"timestamp"`
+}
+
+type deliveryDelay struct {
+	Timestamp time.Time `json:"timestamp"`
+}
+
+func handler(ctx context.Context, snsEvent events.SNSEvent) {
+	// Process the SES delivery notifications included in the SNS event.
+	// An event may include notifications one or multiple recipients
+	// A single recipient might receive multiple notifications
+	// https://docs.aws.amazon.com/ses/latest/dg/notification-contents.html
+	for _, record := range snsEvent.Records {
+		var notification sesNotification
+		err := json.Unmarshal([]byte(record.SNS.Message), &notification)
+		if err != nil {
+			logger.Error().Err(err).Msg("failed to unmarshal notification")
+			continue
+		}
+
+		// Check the status type and get the associated timestamp
+		// Bounce, Complaint, Delivery, Send, Reject, Open, Click, Rendering Failure, DeliveryDelay, or Subscription.
+		var statusTimestamp time.Time
+		switch notification.EventType {
+		case "Bounce":
+			statusTimestamp = notification.Bounce.Timestamp
+		case "Complaint":
+			statusTimestamp = notification.Complaint.Timestamp
+		case "Delivery":
+			statusTimestamp = notification.Delivery.Timestamp
+		case "Send":
+		case "Reject":
+		case "Open":
+			continue // Skip "Open" events
+		case "Click":
+			continue // Skip "Click" events
+		case "Rendering Failure":
+		case "DeliveryDelay":
+			statusTimestamp = notification.DeliveryDelay.Timestamp
+		case "Subscription":
+		default:
+			logger.Warn().Msg("unknown event type " + notification.EventType)
+		}
+
+		// Get Idempotency key from tags to use as partition key, skip if it is not present
+		var idempotencyKey string
+		for _, tag := range notification.Mail.Tags {
+			if *(tag.Name) == "idempotencyKey" {
```

Comment:
Could this lead to a nil pointer, there is a helper `ptr.String(tag.Name)` or `ptr.StringOr(tag.Name, "your value")` which does a nil check and conversion